### ISSUE TITLE
Under/overlinking warnings not quite correct

### ISSUE THREAD

**Issue by wolfv** [prefix-dev/rattler-build#590]
I am building a more complicated package (poppler) on Linux and I get these warnings.

Recipe can be found here: https://gist.github.com/wolfv/a69247c2b583358fc78643335913c4a2

```
2024-02-02T20:57:35.991128Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991171Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991175Z  WARN rattler_build::post_process: Underlinking against libglib
2024-02-02T20:57:35.991178Z  WARN rattler_build::post_process: Underlinking against freetype
2024-02-02T20:57:35.991196Z  WARN rattler_build::post_process: Underlinking against libglib
2024-02-02T20:57:35.991201Z  WARN rattler_build::post_process: Underlinking against cairo
2024-02-02T20:57:35.991206Z  WARN rattler_build::post_process: Underlinking against libglib
2024-02-02T20:57:35.991208Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991224Z  WARN rattler_build::post_process: Underlinking against fontconfig
2024-02-02T20:57:35.991228Z  WARN rattler_build::post_process: Underlinking against libtiff
2024-02-02T20:57:35.991230Z  WARN rattler_build::post_process: Underlinking against openjpeg
2024-02-02T20:57:35.991233Z  WARN rattler_build::post_process: Underlinking against freetype
2024-02-02T20:57:35.991249Z  WARN rattler_build::post_process: Underlinking against libjpeg-turbo
2024-02-02T20:57:35.991267Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991287Z  WARN rattler_build::post_process: Underlinking against libzlib
2024-02-02T20:57:35.991307Z  WARN rattler_build::post_process: Underlinking against lcms2
2024-02-02T20:57:35.991312Z  WARN rattler_build::post_process: Underlinking against libpng
2024-02-02T20:57:35.991315Z  WARN rattler_build::post_process: Underlinking against libglib
2024-02-02T20:57:35.991335Z  WARN rattler_build::post_process: Underlinking against cairo
2024-02-02T20:57:35.991353Z  WARN rattler_build::post_process: Underlinking against freetype
2024-02-02T20:57:35.991371Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991375Z  WARN rattler_build::post_process: Underlinking against libglib
2024-02-02T20:57:35.991378Z  WARN rattler_build::post_process: Underlinking against libglib
2024-02-02T20:57:35.991380Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991382Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991384Z  WARN rattler_build::post_process: Underlinking against libiconv
2024-02-02T20:57:35.991386Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991389Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991392Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991396Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991400Z  WARN rattler_build::post_process: Underlinking against lcms2
2024-02-02T20:57:35.991402Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991404Z  WARN rattler_build::post_process: Underlinking against libiconv
2024-02-02T20:57:35.991406Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991408Z  WARN rattler_build::post_process: Underlinking against lcms2
2024-02-02T20:57:35.991410Z  WARN rattler_build::post_process: Underlinking against libpng
2024-02-02T20:57:35.991426Z  WARN rattler_build::post_process: Underlinking against libtiff
2024-02-02T20:57:35.991430Z  WARN rattler_build::post_process: Underlinking against libzlib
2024-02-02T20:57:35.991433Z  WARN rattler_build::post_process: Underlinking against fontconfig
2024-02-02T20:57:35.991436Z  WARN rattler_build::post_process: Underlinking against openjpeg
2024-02-02T20:57:35.991459Z  WARN rattler_build::post_process: Underlinking against freetype
2024-02-02T20:57:35.991482Z  WARN rattler_build::post_process: Underlinking against libjpeg-turbo
2024-02-02T20:57:35.991486Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991488Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991491Z  WARN rattler_build::post_process: Underlinking against libglib
2024-02-02T20:57:35.991493Z  WARN rattler_build::post_process: Underlinking against libglib
2024-02-02T20:57:35.991495Z  WARN rattler_build::post_process: Underlinking against cairo
2024-02-02T20:57:35.991497Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991501Z  WARN rattler_build::post_process: Underlinking against freetype
2024-02-02T20:57:35.991503Z  WARN rattler_build::post_process: Underlinking against libglib
2024-02-02T20:57:35.991505Z  WARN rattler_build::post_process: Underlinking against freetype
2024-02-02T20:57:35.991509Z  WARN rattler_build::post_process: Underlinking against lcms2
2024-02-02T20:57:35.991511Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991513Z  WARN rattler_build::post_process: Underlinking against cairo
2024-02-02T20:57:35.991516Z  WARN rattler_build::post_process: Underlinking against freetype
2024-02-02T20:57:35.991519Z  WARN rattler_build::post_process: Underlinking against lcms2
2024-02-02T20:57:35.991538Z  WARN rattler_build::post_process: Underlinking against fontconfig
2024-02-02T20:57:35.991542Z  WARN rattler_build::post_process: Underlinking against libjpeg-turbo
2024-02-02T20:57:35.991544Z  WARN rattler_build::post_process: Underlinking against libzlib
2024-02-02T20:57:35.991547Z  WARN rattler_build::post_process: Underlinking against libpng
2024-02-02T20:57:35.991563Z  WARN rattler_build::post_process: Underlinking against openjpeg
2024-02-02T20:57:35.991568Z  WARN rattler_build::post_process: Underlinking against libtiff
2024-02-02T20:57:35.991571Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991573Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991576Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991579Z  WARN rattler_build::post_process: Underlinking against lcms2
2024-02-02T20:57:35.991581Z  WARN rattler_build::post_process: Underlinking against libstdcxx-ng
2024-02-02T20:57:35.991584Z  WARN rattler_build::post_process: Underlinking against libiconv
Relink done!
```

However, when I look at one of the shared libraries of `poppler` I see:

```
wolfv@DESKTOP-C9HHNIV:~/Programs/poppler$ ldd ~/.cache/rattler/cache/pkgs/poppler-24.01.0-hb0f4dca_0/lib/libpoppler.so
        linux-vdso.so.1 (0x00007fffec1ba000)
        libfreetype.so.6 => /lib/x86_64-linux-gnu/libfreetype.so.6 (0x00007f7e8dd88000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f7e8dd6c000)
        libfontconfig.so.1 => /lib/x86_64-linux-gnu/libfontconfig.so.1 (0x00007f7e8dd22000)
        libjpeg.so.8 => /lib/x86_64-linux-gnu/libjpeg.so.8 (0x00007f7e8dca1000)
        libopenjp2.so.7 => not found
        liblcms2.so.2 => not found
        libpng16.so.16 => /lib/x86_64-linux-gnu/libpng16.so.16 (0x00007f7e8dc64000)
        libtiff.so.6 => not found
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f7e8dc5f000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f7e8da33000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f7e8d94c000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f7e8d721000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f7e8e1f6000)
        libbrotlidec.so.1 => /lib/x86_64-linux-gnu/libbrotlidec.so.1 (0x00007f7e8d713000)
        libexpat.so.1 => /lib/x86_64-linux-gnu/libexpat.so.1 (0x00007f7e8d6e2000)
        libuuid.so.1 => /lib/x86_64-linux-gnu/libuuid.so.1 (0x00007f7e8d6d9000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f7e8d6b9000)
        libbrotlicommon.so.1 => /lib/x86_64-linux-gnu/libbrotlicommon.so.1 (0x00007f7e8d694000)
```

So I think our logic is a bit flawed here in some way.

**Comment by orhun** [comment]
Currently we only check the `run` dependencies and since there is no run dependencies it thinks that everything is underlinking. Maybe we should check `run` and `host` together?

Also, there are some false positives due to prefixes/suffixes. e.g. `glib` != `libglib`

I'm also fixing those and working on other improvements.

**Comment by orhun** [comment]
I decided to use the resolved run dependencies which solves the both issues I think.

```rust
    let run_dependencies: Vec<String> = output
        .finalized_dependencies
        .clone()
        .unwrap()
        .run
        .depends
        .iter()
        .flat_map(|v| v.spec().name.to_owned().map(|v| v.as_source().to_owned()))
        .collect();
```

**Comment by orhun** [comment]
Here is the result with the recent fixes:

```
[lib/libpoppler-glib.so.8.26.0] links against:
 ├─ libgio-2.0.so.0 (libglib)
 ├─ libglib-2.0.so.0 (libglib)
 ├─ libc.so.6 (system)
 ├─ ld-linux-x86-64.so.2 (system)
 ├─ libstdc++.so.6 (libstdcxx-ng)
 ├─ libm.so.6 (system)
 ├─ libgobject-2.0.so.0 (libglib)
 ├─ libcairo.so.2 (cairo)
 ├─ libfreetype.so.6 (freetype)
 └─ libpoppler.so.133 (package)


[bin/pdfseparate] links against:
 ├─ libc.so.6 (system)
 ├─ libpoppler.so.133 (package)
 └─ libstdc++.so.6 (libstdcxx-ng)


[lib/libpoppler-cpp.so.0.11.0] links against:
 ├─ libiconv.so.2 (libiconv)
 ├─ libstdc++.so.6 (libstdcxx-ng)
 ├─ libc.so.6 (system)
 └─ libpoppler.so.133 (package)


[lib/libpoppler.so.133.0.0] links against:
 ├─ ld-linux-x86-64.so.2 (system)
 ├─ libtiff.so.6 (libtiff)
 ├─ libc.so.6 (system)
 ├─ libopenjp2.so.7 (openjpeg)
 ├─ libpthread.so.0 (system)
 ├─ libm.so.6 (system)
 ├─ liblcms2.so.2 (lcms2)
 ├─ libfreetype.so.6 (freetype)
 ├─ libz.so.1 (libzlib)
 ├─ libjpeg.so.8 (libjpeg-turbo)
 ├─ libpng16.so.16 (libpng)
 ├─ libstdc++.so.6 (libstdcxx-ng)
 └─ libfontconfig.so.1 (fontconfig)


[bin/pdftocairo] links against:
 ├─ ld-linux-x86-64.so.2 (system)
 ├─ libpoppler.so.133 (package)
 ├─ libstdc++.so.6 (libstdcxx-ng)
 ├─ libm.so.6 (system)
 ├─ libcairo.so.2 (cairo)
 ├─ libfreetype.so.6 (freetype)
 ├─ libc.so.6 (system)
 └─ liblcms2.so.2 (lcms2)


[bin/pdftoppm] links against:
 ├─ libpoppler.so.133 (package)
 ├─ libm.so.6 (system)
 ├─ libc.so.6 (system)
 ├─ liblcms2.so.2 (lcms2)
 └─ libstdc++.so.6 (libstdcxx-ng)


[bin/pdfattach] links against:
 ├─ libpoppler.so.133 (package)
 ├─ libc.so.6 (system)
 └─ libstdc++.so.6 (libstdcxx-ng)


[bin/pdfimages] links against:
 ├─ libpoppler.so.133 (package)
 ├─ libstdc++.so.6 (libstdcxx-ng)
 └─ libc.so.6 (system)


[bin/pdftops] links against:
 ├─ libpoppler.so.133 (package)
 ├─ liblcms2.so.2 (lcms2)
 ├─ libstdc++.so.6 (libstdcxx-ng)
 └─ libc.so.6 (system)


[bin/pdftotext] links against:
 ├─ libc.so.6 (system)
 ├─ libpoppler.so.133 (package)
 └─ libstdc++.so.6 (libstdcxx-ng)


[bin/pdfdetach] links against:
 ├─ libstdc++.so.6 (libstdcxx-ng)
 ├─ libpoppler.so.133 (package)
 └─ libc.so.6 (system)


[bin/pdfinfo] links against:
 ├─ libstdc++.so.6 (libstdcxx-ng)
 ├─ libpoppler.so.133 (package)
 └─ libc.so.6 (system)


[bin/pdfunite] links against:
 ├─ libstdc++.so.6 (libstdcxx-ng)
 ├─ libpoppler.so.133 (package)
 └─ libc.so.6 (system)


[bin/pdffonts] links against:
 ├─ libstdc++.so.6 (libstdcxx-ng)
 ├─ libpoppler.so.133 (package)
 └─ libc.so.6 (system)


[bin/pdftohtml] links against:
 ├─ libm.so.6 (system)
 ├─ libc.so.6 (system)
 ├─ libstdc++.so.6 (libstdcxx-ng)
 └─ libpoppler.so.133 (package)

2024-02-12T09:35:46.184544Z  WARN rattler_build::post_process::checks: Overdepending against fonts-conda-ecosystem
2024-02-12T09:35:46.184574Z  WARN rattler_build::post_process::checks: Overdepending against gettext
2024-02-12T09:35:46.184600Z  WARN rattler_build::post_process::checks: Overdepending against libcurl
```

Regarding the overdepending warnings:

- `fonts-conda-ecosystem`: this dependency comes from `fontconfig` and `fontconfig` is linked. So in this case we should not warn about this I think.
-  `gettext`: this is not linked but it is in the host dependencies. (actual overdepending?)
- `libcurl`: same case as `gettext`

**Comment by wolfv** [comment]
`fonts-conda-ecosystem` "PackageNature" is probably also not a "DSO" package, right? I think we can ignore any non-DSO packages in this check.

And yes, it looks like you are right that curl and gettext in this example are proper "overdepending" issues. You could try removing them - the build might still worK :)

**Comment by orhun** [comment]
Yup, it is now fixed in #619

Also `gettext` and `libcurl` were actually overdepending - I removed them from host and the build worked.
